### ISSUE TITLE
Using Docker to streamline Setup

### ISSUE THREAD

**Issue by gideonite** [cBioPortal/cbioportal#3]
Docker is a tool for helping developers create images to be deployed anywhere: in the cloud, on your in house server, or on your personal machine. There are already docker base images for MySQL and Apache Tomcat and I think that we may be able to build on those images to create a cBioPortal database image, a cBioPortal server image, etc.

I think that Docker might allow people to avoid installing MySQL and importing data. This might reduce the number of problems that people have as well as lower the bar of entry for new users.

Some links from my little bit of research:
- https://www.docker.com/
- on managing data within containers (e.g. for MySQL): http://docs.docker.com/userguide/dockervolumes/
- MySQL image: https://registry.hub.docker.com/_/mysql/
- Tomcat image: https://registry.hub.docker.com/_/tomcat/

**Comment by n1zea144** [comment]
Yeah, its been on our todo list for a while.  Although for various legal reasons, we've decided not to distributed precompiled versions of our code, so it won't be a complete solution.
B

**Comment by gideonite** [comment]
That's too bad. What about a partial solution then? Or is setting up the MySQL database considered to be simple?

**Comment by n1zea144** [comment]
I think a docker image which has a seed database/tomcat ready to go can save some time.  Maybe it could even have a tagged version of the relevant github projects?

**Comment by tjrivera** [comment]
Hi all, 

I've been working on a cBio instance at CHOP that uses a Docker containers to do cBio builds. The process looks something like this:

```bash
#!/usr/bin/env bash

HOME_DIR=~
eval HOME_DIR=$HOME_DIR
LOCAL_MAVEN_REPO="$HOME_DIR/.cbio_build/"

mkdir -p $LOCAL_MAVEN_REPO

echo "Starting MySQL instance..."
docker run -d --name mysql_test -e MYSQL_ROOT_PASSWORD=root123 mysql:5.5
sleep 10
echo "Creating base DBs"
docker run -it\
    --rm\
    --link mysql_test:db\
    mysql:5.5 mysql -h db -u root --password=root123 -e "create database cbioportal; create database cgds_test;"
echo "Building cBio"
docker run\
    -it\
    --rm\
    --link mysql_test:db\
    --name cbioportal_build\
    -e PORTAL_HOME=..\
    -v $LOCAL_MAVEN_REPO:/opt/data/maven-repo\
    -v "$PWD":/usr/src/mymaven\
    -w /usr/src/mymaven\
    maven:3.2-jdk-7 mvn clean install -Dmaven.test.skip=true -Dmaven.repo.local=/opt/data/maven-repo
echo "Removing Build DB"
docker stop mysql_test && docker rm mysql_test
docker build -t cbttc_cbio .
```

I then drop the resulting war file into a tomcat container and mount in the necessary configuration files. It's been incredibly useful as I'm not a java developer and have very little experience with setting up tomcat or maven locally. I also don't have to worry about setting up MySQL on any build machine. This allows me to drop the cBio code and my script on any VM and be confident that it will build and run. 

Is there any interest in fleshing this out more? I'd be happy to work on a pull-request to move builds into containers. 

This would also solve #14 as the target MySQL db would always be a container.

**Comment by tjrivera** [comment]
For additional detail this is my cBio Dockerfile:

```
FROM tomcat:8.0

ENV PORTAL_HOME /usr/data/cbio/cbio_portal_home

WORKDIR /usr/local/tomcat

ADD portal/target/cbioportal.war /usr/local/tomcat/webapps/cbioportal.war
ADD docker/tomcat-users.xml.sample /usr/local/tomcat/conf/tomcat-users.xml
ADD docker/context.xml /usr/local/tomcat/conf/context.xml
ADD docker/mysql-connector-java-5.1.16.jar /usr/local/tomcat/lib/
ADD src/main/resources /usr/data/cbio/cbio_portal_home

EXPOSE 8080

```

**Comment by tjrivera** [comment]
@n1zea144 any opinion on this?

**Comment by n1zea144** [comment]
Sorry for the delay.  I am interested in this, but I'm still working on clearing my stack.

**Comment by inodb** [comment]
@tjrivera I'm interested in trying this. How does this work exactly? Not super familiar with Docker yet. Can I run the above posted bash file with the Dockerfile in the same directory? I assume I have to keep the mysql server running?

**Comment by tjrivera** [comment]
@inodb, sorry for the late reply. The bash file above would reside in the root of the cbioportal repository alongside the Dockerfile. 

I create a local maven repo on my local machine and mount into the running container at build time so that I can persist maven's packages across subsequent builds. 

cbio needs a MySQL database running during its build. So i start one temporarily, build cbio using the maven container (which results in the war file being dropped in `$(PWD)`, stop the temp mysql db, remove the temp mysql db, and then build the cbio container using the war file that the maven container dropped. 

this totally "contains" the build process -- so i can build cbio on any machine that contains the cbio code and has access to a docker daemon. i don't have to worry about any other build dependencies or outside factors

**Comment by inodb** [comment]
The compilation step is failing for me on 634ecaa6d39fec1b1c154d5350f7581f067fde66.

I'm using boot2docker on osx:

```
Boot2Docker-cli version: v1.5.0
Git commit: ccd9032
```

Any clue?

```
[INFO] Portal Core ........................................ FAILURE [ 19.265 s]
[INFO] Portal Web Layer ................................... SKIPPED
[INFO] cBio Cancer Genomics Portal ........................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 49.577 s
[INFO] Finished at: 2015-08-28T15:53:29+00:00
[INFO] Final Memory: 40M/287M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:sql-maven-plugin:1.5:execute (default) on project core: Communications link failure due to underlying exception:
[ERROR]
[ERROR] ** BEGIN NESTED EXCEPTION **
[ERROR]
[ERROR] java.net.SocketException
[ERROR] MESSAGE: java.net.ConnectException: Connection refused
[ERROR]
[ERROR] STACKTRACE:
[ERROR]
[ERROR] java.net.SocketException: java.net.ConnectException: Connection refused
[ERROR] at com.mysql.jdbc.StandardSocketFactory.connect(StandardSocketFactory.java:156)
[ERROR] at com.mysql.jdbc.MysqlIO.<init>(MysqlIO.java:276)
[ERROR] at com.mysql.jdbc.Connection.createNewIO(Connection.java:2641)
[ERROR] at com.mysql.jdbc.Connection.<init>(Connection.java:1531)
[ERROR] at com.mysql.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:266)
[ERROR] at org.codehaus.mojo.sql.SqlExecMojo.getConnection(SqlExecMojo.java:899)
[ERROR] at org.codehaus.mojo.sql.SqlExecMojo.execute(SqlExecMojo.java:612)
[ERROR] at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
[ERROR] at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
[ERROR] at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
[ERROR] at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
[ERROR] at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
[ERROR] at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
[ERROR] at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
[ERROR] at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
[ERROR] at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
[ERROR] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] at java.lang.reflect.Method.invoke(Method.java:606)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR] at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR]
[ERROR]
[ERROR] ** END NESTED EXCEPTION **
```

**Comment by tjrivera** [comment]
Ah, I left out a key detail: you have to change the connection settings to the test db [here] (https://github.com/cBioPortal/cbioportal/blob/bb3ace1a6b4148a8dd46c275660bf88b88097c43/core/pom.xml#L359) to `jdbc:mysql://db:3306/cgds_test`

**Comment by inodb** [comment]
Getting further!
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 24.808 s
[INFO] Finished at: 2015-08-28T16:29:15+00:00
[INFO] Final Memory: 41M/355M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:sql-maven-plugin:1.5:execute (default) on project core: Access denied for user ''@'172.17.0.9' (using password: NO) -> [Help 1]
```
Should I change the properties file or smth?

**Comment by tjrivera** [comment]
My pom.xml includes:

```xml
<settingsKey>settingsKey</settingsKey>
<driver>com.mysql.jdbc.Driver</driver>
<url>jdbc:mysql://db:3306/cgds_test</url>
<username>root</username>
<password>root123</password>
```

**Comment by inodb** [comment]
Compilation works now, thanks! How do I run the image?

I tried
```
docker run -it -P --rm cbttc_cbio
```
and some various `-p` and `-P` flags, but can't seem to ping the 8080 port

**Comment by inodb** [comment]
Problem was related to boot2docker not exposing the ports of the vm to the host machine. I just installed the latest [Docker Toolbox](https://www.docker.com/toolbox) and now it works fine. Thanks for the help!

**Comment by alexsigaras** [comment]
This issue has been resolved in release-1.5.0
